### PR TITLE
Mae/api for new structure

### PR DESCRIPTION
**Pull request by hdphuong**
Allow users to have multiple majors and **minors** (we can't really enforce a restriction on number here since that info is not on the handbook)
- a new data structure for programs (thanks @Joelrwest for making `programs_processing.py` more readable)
- APIs for the new data structure
- fill the `General` tab on the FE
- fixed search bug

### CODE REVIEW

**Comment by gtangelo** [comment]
Note: Update the `redux-persist` data version and check if the progression checker is not broken as a result of these changes :)

**Review by joooooooooooooooooooooooooooooooooooosh — APPROVED** [review]

File: frontend/src/store.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ const rootReducer = combineReducers({
 
 // migration schema used to translate structure - return undefined to reset initialState
 const migrations = {
-  0: () => undefined,
```
Comment (gtangelo) [inline frontend/src/store.js:40]: We should keep this back in

**Review by gtangelo — APPROVED** [review]
Just need to address that comment i made. Otherwise, Lgtm for the FE side!!

File: frontend/src/store.js (hunk 1 above)
Comment (hdphuong) [reply frontend/src/store.js:40]: should it be 1 or 0? 😮

Comment (gtangelo) [reply frontend/src/store.js:40]: Both! Ideally u should append new key-value pairing to the object rather than overriding it? If thAt makes sense haha

Comment (hdphuong) [reply frontend/src/store.js:40]: ah icic thanks for picking this up 😃

File: backend/server/routers/programs.py [hunk 2]
```diff
@@ -158,11 +160,26 @@ def convertSubgroupObjectToCoursesDict(object: str, description: str|list[str]):
...
+    #TODO: further standardise non_spec_data to remove these line:
+    title = container.get("title")
+    if container.get("type") == "gened":
+        title = "General Education"
+    
+    # don't do anything if it's info_rule or limit_rule
+    if container.get("type") is not None and "rule" in container.get("type"):
+        return []
+
+    if container.get("credits_to_complete") is None:
```
Comment (Joelrwest) [inline backend/server/routers/programs.py]: It's impossible for it to be None anymore, it'll be set to 0 if there's some error

File: backend/server/routers/programs.py [hunk 3]
```diff
@@ -158,11 +160,26 @@ def convertSubgroupObjectToCoursesDict(object: str, description: str|list[str]):
...
-    item = structure[type][container["title"]]
-
+    #TODO: further standardise non_spec_data to remove these line:
+    title = container.get("title")
+    if container.get("type") == "gened":
+        title = "General Education"
+    
+    # don't do anything if it's info_rule or limit_rule
+    if container.get("type") is not None and "rule" in container.get("type"):
+        return []
```
Comment (Joelrwest) [inline backend/server/routers/programs.py:168]: We going to do anything in the future with it? Some of the limit rules are important, and some of the information rules just seem like something good to display to the user (they're just in plain English so there's no shot we'll format them)

File: backend/server/routers/programs.py (hunk 2 above)
Comment (Joelrwest) [reply backend/server/routers/programs.py]: (can change if wanted)

File: backend/server/routers/programs.py [hunk 4]
```diff
@@ -268,35 +285,24 @@ def getStructure(
...
     if minor:
-        addSpecialisation(structure, minor, "Minor")
+        minors_l = minor.split("+") if "+" in minor else [minor]
+        for m in minors_l:
+            addSpecialisation(structure, m, "Minor")
 
     # add details for program code
     programsResult = programsCOL.find_one({"code": programCode})
     if not programsResult:
+        print(programCode)
```
Comment (Joelrwest) [inline backend/server/routers/programs.py:294]: Debug print?

File: backend/server/tests/programs/test_getMinors.py [hunk 5]
```diff
@@ -6,9 +6,10 @@ def test_sanity():
     x = requests.get('http://127.0.0.1:8000/programs/getMinors/3778')
     assert x.status_code == 200
     minors = x.json()['minors']
-    for key, value in minors.items():
-        assert re.match(r"[A-Z]{5}[0-9]{1}", key)
-        assert type(value) is str
+    for minor in minors.values():
+        for key, value in minor["specs"].items():
+            assert re.match(r"[A-Z]{5}[0-9]{1}", key)
+            assert type(value) is str
```
Comment (Joelrwest) [inline backend/server/tests/programs/test_getMinors.py]: Mega nitpick but use isinstance(value, str)

**Review by Joelrwest — APPROVED** [review]

**Comment by Joelrwest** [comment]
My changes are (almost) all to do with programs_processing.py which processes the data from programsFormattedRaw.json and outputs a processed version to programsProcessed.json

What I've done:
- Made programs_processing.py (hopefully) massively more readable
- Notes are now available for each section. E.g., in CS for the major the note "COMPA1 is the default stream, and will......."
- Locations of majors/minors/honours now consistent
- Biggest issue is the 'other' requirements (basically lumping all of limit_rule and non_spec_data into one bucket).
- The types vary massively. It currently deals with core courses, electives with a list of courses, and electives with a list of faculties
- For core courses, I've tried to mimic the way that they're represented in specialisationsProcessed.json in the hopes that we can reuse functionality. An I had were, what is levels?

(Simple) example of changes for normal CS degree (see programsProcessed.json for more examples)
Before:
```
"3778": {
    "title": "Computer Science",
    "code": "3778",
    "duration": 3,
    "UOC": 144,
    "faculty": "Faculty of Engineering",
    "components": {
        "GE": {
            "credits_to_complete": 12
        },
        "FE": {
            "credits_to_complete": 36,
            "Minors": {
                "INFSA2": "Information Systems",
                "FINSA2": "Finance",
                "ACCTA2": "Accounting",
                "PSYCM2": "Psychology",
                "MATHC2": "Mathematics",
                "MARKA2": "Marketing"
            }
        },
        "SpecialisationData": {
            "Majors": {
                "COMPA1": "Computer Science",
                "COMPY1": "Computer Science (Security Engineering)",
                "COMPN1": "Computer Science (Computer Networks)",
                "COMPS1": "Computer Science (Embedded Systems)",
                "COMPJ1": "Computer Science (Programming Languages)",
                "COMPI1": "Computer Science (Artificial Intelligence)",
                "COMPE1": "Computer Science (eCommerce Systems)",
                "COMPD1": "Computer Science (Database Systems)"
            }
        },
        "NonSpecialisationData": {}
    }
}
```

After:
```
"3778": {
        "title": "Computer Science",
        "code": "3778",
        "duration": 3,
        "UOC": 144,
        "faculty": "Faculty of Engineering",
        "overview": "Computer Science involves the study of the design, construction and <more stuff>",
        "structure_summary": "Students must complete 144 UOC when taken as a standalone program. <more more stuff>",
        "components": {
            "non_spec_data": [
                {
                    "type": "gened",
                    "credits_to_complete": 12,
                    "notes": "Students must take 12 UOC of any General Education course.\nNOTE: <stuff>"
                },
                {
                    "type": "limit_rule",
                    "courses": {
                        "....1": "any level 1 course"
                    },
                    "title": "Maximum Level 1 Electives UOC",
                    "credits_to_complete": 60,
                    "levels": [
                        1
                    ],
                    "notes": "Students may only undertake a maximum of 60 UOC of the following courses."
                }
            ],
            "spec_data": {
                "majors": {
                    "Computer Science": {
                        "is_optional": false,
                        "specs": {
                            "COMPA1": "Computer Science",
                            "COMPY1": "Computer Science (Security Engineering)",
                            "COMPN1": "Computer Science (Computer Networks)",
                            "COMPS1": "Computer Science (Embedded Systems)",
                            "COMPJ1": "Computer Science (Programming Languages)",
                            "COMPI1": "Computer Science (Artificial Intelligence)",
                            "COMPE1": "Computer Science (eCommerce Systems)",
                            "COMPD1": "Computer Science (Database Systems)"
                        },
                        "notes": "COMPA1 is the default stream, and will be used if no other stream is selected."
                    }
                },
                "minors": {
                    "Computer Science": {
                        "is_optional": true,
                        "specs": {
                            "INFSA2": "Information Systems",
                            "FINSA2": "Finance",
                            "ACCTA2": "Accounting",
                            "PSYCM2": "Psychology",
                            "MATHC2": "Mathematics",
                            "MARKA2": "Marketing"
                        },
                        "notes": "Optional minors available include the following. If you <more stuff>"
                    }
                }
            }
        },
        "processing_warnings": []
    }
```

Future:
- Some programs don't have a free elective section, so I've completely ditched it. Instead, we're just going to have to make some system to figure out how many free electives people have
- How to deal with limit rules
- How to display information rules to the user (as well as notes, program descriptions etc)
- Weird rules that seem to contradict themselves e.g., [Science (3970)](https://www.handbook.unsw.edu.au/undergraduate/programs/2022/3970) in the section titled 'Science Elective Courses' and 'Science Electives' unless I'm just reading them wrong
- How to deal with course substitutions (e.g., in [BE/CS (3785)](https://www.handbook.unsw.edu.au/undergraduate/programs/2022/3785?year=2022) you can sub 2nd year maths courses for MATH1081)
- Ongoing issues where programsProcessed.json is missing some courses/specialisations because of different formatting by the handbook that I haven't seen before. These are usually very obvious

File: backend/server/routers/programs.py (hunk 3 above)
Comment (hdphuong) [reply backend/server/routers/programs.py:168]: yup we don't have a place to display them yet but i agree we should. will ask the FE team to do that after this is merged

File: backend/server/routers/programs.py (hunk 2 above)
Comment (hdphuong) [reply backend/server/routers/programs.py]: yup, changed to 0 by default
